File: web/src/random.ts
export function randString(charset: string, minLen: number, maxLen: number) {
    const length = Math.floor(Math.random() * (maxLen - minLen + 1)) + minLen;
    const arr = new Uint8Array(length);
    crypto.getRandomValues(arr);
    let string = '';

    for (let i = 0; i < length; i++) {
        string += charset[arr[i] % charset.length];
    }

    return string;
}

function randInt(min: number, max: number): number {
    const rand = Math.floor(Math.random() * (max - min + 1)) + min;
    return rand;
}

export function randSubdomain() {
    const charset = "abcdefghijklmnopqrstuvwxyz0123456789-";
    let subdomain: string;
    
    do {
        subdomain = randString(charset, 16, 32);
    } while (subdomain.startsWith('-') || subdomain.endsWith('-'));
    
    return subdomain;
}

export function randCode(): string {
    const minVars = 50, maxVars = 500;
    const minFuncs = 50, maxFuncs = 500;
    const charset = 'abcdefghijklmnopqrstuvwxyz0123456789';

    const varCount = randInt(minVars, maxVars);
    const funcCount = randInt(minFuncs, maxFuncs);

    let varsBuilder = '';
    for (let i = 0; i < varCount; i++) {
        const varName = `__var_${randString(charset, 8, 16)}_${i}`;
        const value = randInt(0, 99999);
        varsBuilder += `let ${varName} = ${value};\n`;
    }

    let funcsBuilder = '';
    for (let i = 0; i < funcCount; i++) {
        const funcName = `__func_${randString(charset, 8, 16)}_${i}`;
        const value = randInt(0, 999);
        funcsBuilder += `function ${funcName}() { return ${value}; }\n`;
    }

    return varsBuilder + funcsBuilder;
}

File: internal/random.go
package internal

import (
	"fmt"
	mrand "math/rand"
	"strings"
)

func randInt(min, max int) int {
	return mrand.Intn(max-min+1) + min
}

func randString(charset string, minLen, maxLen int) string {
	length := minLen + mrand.Intn(maxLen-minLen+1)
	bytes := make([]byte, length)
	for i := range bytes {
		bytes[i] = charset[mrand.Intn(len(charset))]
	}
	
	return string(bytes)
}

func randSubdomain() (string, error) {
	const charset = "abcdefghijklmnopqrstuvwxyz0123456789-"
	for {
		subdomain := randString(charset, 16, 32)
		if !strings.HasPrefix(subdomain, "-") && !strings.HasSuffix(subdomain, "-") {
			return subdomain, nil
		}
	}
}

func randCode() string {
	const minVars, maxVars = 50, 500
	const minFuncs, maxFuncs = 50, 500
	const charset = "abcdefghijklmnopqrstuvwxyz0123456789"
	varCount := randInt(minVars, maxVars)
	funcCount := randInt(minFuncs, maxFuncs)

	var varsBuilder strings.Builder
	for i := range varCount {
		varName := fmt.Sprintf("__var_%s_%d", randString(charset, 8,16), i)
		value := randInt(0, 99999)
		varsBuilder.WriteString(fmt.Sprintf("let %s = %d;\n", varName, value))
	}

	var funcsBuilder strings.Builder
	for i := range funcCount {
		funcName := fmt.Sprintf("__func_%s_%d", randString(charset, 8,16), i)
		value := randInt(0, 999)
		fmt.Fprintf(&funcsBuilder, "function %s() { return %d; }\n", funcName, value)
	}

	return varsBuilder.String() + funcsBuilder.String()
}
